MAINT: reduce void type repr code duplication

tylerjereddy opened the pull request:
The void type `__str__` and `__repr__` handling code at the C level was almost completely identical in their respective C functions. Both already called common `__repr__` handling code at the Python level, and this PR aims to similarly reduce code duplication / maintenance burden at the C level.

Note that it is, of course, not uncommon for `__str__` implementations to simply call `__repr__`, and the quaternion dtype implementation uses [exact duplicate code for both](https://github.com/moble/quaternion/blob/master/numpy_quaternion.c#L760) instead of one calling the other.

Here, both functions weren't quite identical, so the common handler function accepts arguments to allow the formatting differences between `__repr__` and `__str__` to be specified while abstracting the common logic.

void type discussion / work in PR from @ahaldane : https://github.com/numpy/numpy/pull/8981

eric-wieser commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -550,27 +550,21 @@ voidtype_repr(PyObject *self)
+    return voidtype_repr_handler(self, "void(b'", "')");
+}
 
-        return PyObject_CallFunction(reprfunc, "O", self);
```
I'd maybe be inclined just to extract the contents of this `if` statement into a helper function - we might want to have different behavior for `str` and `repr` of a structured void in future.

eric-wieser replied:
And then
```
static PyObject *
voidtype_str(PyObject *self)
{
    PyVoidScalarObject *s = (PyVoidScalarObject*) self;
    if (PyDataType_HASFIELDS(s->descr)) {
        return _void_scalar_repr(self);
    }
    return _void_to_hex(s->obval, s->descr->elsize, "b'", "\\x", "'");
}
```

I'd prefer for every void member function to have an `if (PyDataType_HASFIELDS ...)` member to reinforce the idea that  structured types and raw-byte types should be two separate things

tylerjereddy committed:
MAINT: reduce void type repr code duplication

* voidtype_str and voidtype_repr previously
had nearly-identical code; the shared code
has now been abstracted to a handler function

tylerjereddy commented:
Thanks Eric, revised accordingly (I hope).
